service-statistics/src/main/java/org/oskari/statistics/user/StatisticalIndicatorServiceMybatisImpl.java

```diff
@@ -94,37 +98,110 @@ public void deleteByUser(long userId) {
-        ind.setId(String.valueOf(userIndicator.getId()));
+        ind.setId(String.valueOf(userIndicator.id));
 
-        JSONObject title = JSONHelper.createJSONObject(userIndicator.getTitle());
+        JSONObject title = JSONHelper.createJSONObject(userIndicator.title);
         Iterator<String> langKeys = title.keys();
         while (langKeys.hasNext()) {
             String lang = langKeys.next();
             ind.addName(lang, title.optString(lang));
         }
 
-        JSONObject source = JSONHelper.createJSONObject(userIndicator.getSource());
+        JSONObject source = JSONHelper.createJSONObject(userIndicator.source);
         langKeys = source.keys();
         while (langKeys.hasNext()) {
             String lang = langKeys.next();
             ind.addSource(lang, source.optString(lang));
         }
 
-        JSONObject desc = JSONHelper.createJSONObject(userIndicator.getDescription());
+        JSONObject desc = JSONHelper.createJSONObject(userIndicator.description);
         langKeys = desc.keys();
         while (langKeys.hasNext()) {
             String lang = langKeys.next();
-            ind.addSource(lang, desc.optString(lang));
+            ind.addDescription(lang, desc.optString(lang));
         }
-        ind.setPublic(userIndicator.isPublished());
-        ind.addLayer(new StatisticalIndicatorLayer(userIndicator.getMaterial(), ind.getId()));
+        ind.setPublic(userIndicator.published);
 
-        // If we want to provide year, need to do it like this. But currently there's always just one choice
+        // Initialize the year dimension as the only one.
         StatisticalIndicatorDataDimension dim = new StatisticalIndicatorDataDimension("year");
-        dim.addAllowedValue(String.valueOf(userIndicator.getYear()));
         ind.getDataModel().addDimension(dim);
         return ind;
     }
+
+    private UserIndicatorDataRow createDBRow(StatisticalIndicator userIndicator) {
+        UserIndicatorDataRow row = new UserIndicatorDataRow();
+        row.title = new JSONObject(userIndicator.getName()).toString();
+        row.description = new JSONObject(userIndicator.getDescription()).toString();
+        row.source = new JSONObject(userIndicator.getSource()).toString();
+        row.published = userIndicator.isPublic();
+        return row;
+    }
+
+    private void addDimension(StatisticalIndicator ind, UserIndicatorDataRow row) {
+        if(ind.getLayer(row.regionsetId) == null) {
+            ind.addLayer(new StatisticalIndicatorLayer(row.regionsetId, ind.getId()));
+        }
+        ind.getDataModel().getDimension("year").addAllowedValue(String.valueOf(row.year));
+    }
+
+    public StatisticalIndicator saveIndicator(StatisticalIndicator ind, long userId) {
+        int id = ConversionHelper.getInt(ind.getId(), -1);
+        UserIndicatorDataRow row = createDBRow(ind);
+        row.userId = userId;
+        row.id = id;
+        if (id != -1) {
+            // check that it's the users indicator
```

Comment:
I'd argue this check is unnecessary. The `UserIndicatorMapper#updateIndicator(row)` will return 0 if the row-to-update doesn't exist, so this case is covered twice.